pkg/gather/clusterconfig/operators.go

```diff
@@ -92,28 +96,124 @@ func GatherClusterOperators(g *Gatherer, c chan<- gatherResult) {
+func gatherPodContainersLogs(ctx context.Context, client corev1client.CoreV1Interface, pods []*corev1.Pod, bufferSize int64) ([]record.Record, error) {
+	if bufferSize <= 0 {
+		return nil, fmt.Errorf("invalid buffer size %d", bufferSize)
+	}
+
+	// Fetch a list of containers in pods and calculate a log size quota
+	// Total log size must not exceed maxLogsSize multiplied by logCompressionRatio
+	klog.V(2).Infof("Maximum buffer size: %v bytes", bufferSize)
+	buf := bytes.NewBuffer(make([]byte, 0, bufferSize))
+
+	// Fetch previous and current container logs
+	var records []record.Record
+	for _, isPrevious := range []bool{true, false} {
+		for _, pod := range pods {
+			clog := getContainerLogs(ctx, client, pod, isPrevious, buf, bufferSize)
+			if len(clog) > 0 {
+				records = append(records, clog...)
+			}
+		}
+	}
+
+	return records, nil
+}
+
+// getContainerLogs get previous and current log reports for pod containers using the k8s API response
+func getContainerLogs(ctx context.Context, client corev1client.CoreV1Interface, pod *corev1.Pod, isPrevious bool, buf *bytes.Buffer, bufferSize int64) []record.Record {
+	var records []record.Record
+
+	allContainers := append(pod.Spec.InitContainers, pod.Spec.Containers...)
+	for _, c := range allContainers {
+		// only grab previous log if the pod is restarted
+		if isPrevious && !isPodRestarted(pod) {
+			continue
+		}
+
+		logName := fmt.Sprintf("%s_current.log", c.Name)
+		if isPrevious {
+			logName = fmt.Sprintf("%s_previous.log", c.Name)
+		}
+
+		buf.Reset()
+		klog.V(2).Infof("Fetching logs for %s container %s pod in namespace %s (previous: %v).", c.Name, pod.Name, pod.Namespace, isPrevious)
+		// Fetch container logs and continue on error
+		err := fetchPodContainerLog(ctx, client, pod, buf, c.Name, isPrevious, &bufferSize)
+		if err != nil {
+			klog.V(2).Infof("Error: %q", err)
+			continue
+		}
+
+		// Do not record empty records
+		if buf.Len() == 0 {
+			continue
+		}
+		records = append(records, record.Record{Name: fmt.Sprintf("config/pod/%s/logs/%s/%s", pod.Namespace, pod.Name, logName), Item: marshal.Raw{Str: buf.String()}})
+	}
+
+	return records
+}
+
+// clusterOperatorsRecords generates the cluster operator records
+func clusterOperatorsRecords(ctx context.Context, items []configv1.ClusterOperator, dynamicClient dynamic.Interface, discoveryClient discovery.DiscoveryInterface) []record.Record {
 	resVer, _ := getOperatorResourcesVersions(discoveryClient)
-	records := make([]record.Record, 0, len(config.Items))
-	for idx, co := range config.Items {
+	records := make([]record.Record, 0, len(items))
+
+	for _, co := range items {
 		records = append(records, record.Record{
-			Name: fmt.Sprintf("config/clusteroperator/%s", config.Items[idx].Name),
-			Item: record.JSONMarshaller{Object: &config.Items[idx]},
+			Name: fmt.Sprintf("config/clusteroperator/%s", co.Name),
+			Item: record.JSONMarshaller{Object: &co},
```

Comment:
Unless I'm mistaken, which is entirely possible, this will not work as expected. In this context, `co` is a copy of the structure, not a reference to it, so storing a pointer to the copy will result in records being a list of N instances of the last item in `items`. I can totally see how this mistake is easily made because I keep on making it myself and then wondering why and how stuff that shouldn't be gathered ends up in the archive. Only the line above should have been changed. For the object reference, it is necessary to use the old approach with an index.

Reply:
Ivo is right and yes I made this mistake too. Please fix.

Reply:
I wasn't aware of this. Nice catch, I'll change it right now. Thanks, @natiiix